File: packages/telemetry/typescript/src/sdk/memory.test.ts
import { context, propagation, trace } from "@opentelemetry/api"
import { AsyncLocalStorageContextManager } from "@opentelemetry/context-async-hooks"
import { InMemorySpanExporter } from "@opentelemetry/sdk-trace-node"
import { afterEach, describe, expect, it } from "vitest"
import { capture } from "./context.ts"
import { Latitude } from "./init.ts"
import { createMemoryTelemetry } from "./memory.ts"

function attr(span: { attributes: Record<string, unknown> }, name: string) {
  return span.attributes[name]
}

function newLatitude() {
  const exporter = new InMemorySpanExporter()
  const latitude = new Latitude({ apiKey: "test-key", project: "test-project", exporter, disableBatch: true })
  return { exporter, latitude }
}

describe("createMemoryTelemetry", () => {
  afterEach(() => {
    trace.disable()
    context.disable()
    propagation.disable()
  })

  it("emits a create_memory span with identity attributes and no content by default", async () => {
    const { exporter, latitude } = newLatitude()
    const memory = createMemoryTelemetry({ latitude, storeId: "prefs" })

    await memory.create({ recordId: "mem_1", records: [{ content: "likes tea" }] })
    await latitude.flush()

    const span = exporter.getFinishedSpans().find((s) => s.name === "create_memory")
    expect(span).toBeDefined()
    expect(span?.instrumentationScope.name).toBe("so.latitude.instrumentation.memory")
    expect(span ? attr(span, "gen_ai.operation.name") : undefined).toBe("create_memory")
    expect(span ? attr(span, "gen_ai.memory.store.id") : undefined).toBe("prefs")
    expect(span ? attr(span, "gen_ai.memory.record.id") : undefined).toBe("mem_1")
    expect(span ? attr(span, "gen_ai.memory.record.count") : undefined).toBe(1)
    expect(Number.isInteger(span ? attr(span, "gen_ai.memory.record.count") : undefined)).toBe(true)
    expect(span ? attr(span, "gen_ai.memory.records") : undefined).toBeUndefined()

    await latitude.shutdown()
  })

  it("sends record content as a valid JSON array only when captureContent is on, applying redact", async () => {
    const { exporter, latitude } = newLatitude()
    const memory = createMemoryTelemetry({
      latitude,
      storeId: "prefs",
      captureContent: true,
      redact: (records) => records.map((r) => ({ ...r, content: "[redacted]" })),
    })

    await memory.update({
      recordId: "mem_1",
      records: [{ id: "mem_1", content: "likes tea", metadata: { source: "chat" } }],
    })
    await latitude.flush()

    const span = exporter.getFinishedSpans().find((s) => s.name === "update_memory")
    const raw = span ? attr(span, "gen_ai.memory.records") : undefined
    expect(typeof raw).toBe("string")

    const parsed = JSON.parse(raw as string)
    // Contract self-check: the exact shape ingest validates (non-empty array, every element has `content`).
    expect(Array.isArray(parsed)).toBe(true)
    expect(parsed.length).toBeGreaterThan(0)
    expect(parsed.every((r: { content?: unknown }) => "content" in r)).toBe(true)
    expect(parsed[0].content).toBe("[redacted]")

    await latitude.shutdown()
  })

  it("wraps execute, returns its result, and stamps context attributes", async () => {
    const { exporter, latitude } = newLatitude()
    const memory = createMemoryTelemetry({
      latitude,
      storeId: "prefs",
      context: { sessionId: "sess-1", userId: "user-1", project: "test-project" },
    })

    const result = await memory.create({
      recordId: "mem_1",
      records: [{ content: "hi" }],
      execute: async () => ({ ok: true }),
    })

    expect(result).toEqual({ ok: true })
    await latitude.flush()

    const span = exporter.getFinishedSpans().find((s) => s.name === "create_memory")
    expect(span ? attr(span, "session.id") : undefined).toBe("sess-1")
    expect(span ? attr(span, "user.id") : undefined).toBe("user-1")
    expect(span ? attr(span, "latitude.project") : undefined).toBe("test-project")

    await latitude.shutdown()
  })

  it("records exceptions and rethrows when execute fails", async () => {
    const { exporter, latitude } = newLatitude()
    const memory = createMemoryTelemetry({ latitude, storeId: "prefs" })

    await expect(
      memory.upsert({
        recordId: "mem_1",
        execute: async () => {
          throw new Error("boom")
        },
      }),
    ).rejects.toThrow("boom")
    await latitude.flush()

    const span = exporter.getFinishedSpans().find((s) => s.name === "upsert_memory")
    expect(span?.status.message).toBe("boom")
    expect(span?.events.some((e) => e.name === "exception")).toBe(true)

    await latitude.shutdown()
  })

  it("maps search results to records and always sets the returned count", async () => {
    const { exporter, latitude } = newLatitude()
    const memory = createMemoryTelemetry({ latitude, storeId: "prefs" })

    const hits = [
      { id: "mem_1", content: "likes tea", score: 0.9 },
      { id: "mem_2", content: "lives in Barcelona", score: 0.7 },
    ]
    const result = await memory.search({
      query: "preferences",
      execute: async () => hits,
      recordsFromResult: (r) => r,
    })

    expect(result).toBe(hits)
    await latitude.flush()

    const span = exporter.getFinishedSpans().find((s) => s.name === "search_memory")
    expect(span ? attr(span, "gen_ai.memory.query.text") : undefined).toBe("preferences")
    expect(span ? attr(span, "gen_ai.memory.record.count") : undefined).toBe(2)
    // Count of returned records is not content, so it rides even with capture off.
    expect(span ? attr(span, "gen_ai.memory.records") : undefined).toBeUndefined()

    await latitude.shutdown()
  })

  it("treats a delete without recordId as a whole-store wipe", async () => {
    const { exporter, latitude } = newLatitude()
    const memory = createMemoryTelemetry({ latitude })

    await memory.delete({ storeId: "prefs" })
    await memory.createStore({ storeId: "prefs" })
    await latitude.flush()

    const del = exporter.getFinishedSpans().find((s) => s.name === "delete_memory")
    expect(del ? attr(del, "gen_ai.memory.store.id") : undefined).toBe("prefs")
    expect(del ? attr(del, "gen_ai.memory.record.id") : undefined).toBeUndefined()

    const createStore = exporter.getFinishedSpans().find((s) => s.name === "create_memory_store")
    expect(createStore ? attr(createStore, "gen_ai.operation.name") : undefined).toBe("create_memory_store")

    await latitude.shutdown()
  })

  it("inherits latitude context from an enclosing capture()", async () => {
    const cm = new AsyncLocalStorageContextManager()
    cm.enable()
    context.setGlobalContextManager(cm)

    const { exporter, latitude } = newLatitude()
    const memory = createMemoryTelemetry({ latitude, storeId: "prefs" })

    await capture("agent-run", async () => memory.create({ recordId: "mem_1" }), {
      sessionId: "sess-9",
      userId: "user-9",
    })
    await latitude.flush()

    const span = exporter.getFinishedSpans().find((s) => s.name === "create_memory")
    expect(span ? attr(span, "session.id") : undefined).toBe("sess-9")
    expect(span ? attr(span, "user.id") : undefined).toBe("user-9")

    await latitude.shutdown()
  })
})

File: packages/telemetry/python/tests/telemetry/memory_test.py
"""Tests for the memory-operation span helper (create_memory_telemetry)."""

import json
from typing import Any

from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
from opentelemetry.trace import StatusCode

from latitude_telemetry import Latitude, capture, create_memory_telemetry
from latitude_telemetry.constants import SCOPE_LATITUDE


def _new_latitude() -> tuple[InMemorySpanExporter, Latitude]:
    exporter = InMemorySpanExporter()
    lat = Latitude(
        api_key="test-api-key",
        project="test-project",
        disable_batch=True,
        tracer_provider=TracerProvider(),
        exporter=exporter,
    )
    return exporter, lat


def _span(exporter: InMemorySpanExporter, name: str):
    return next((s for s in exporter.get_finished_spans() if s.name == name), None)


class TestMemoryTelemetry:
    def test_emits_create_memory_span_without_content_by_default(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(lat, store_id="prefs")

        memory.create(record_id="mem_1", records=[{"content": "likes tea"}])
        lat.flush()

        span = _span(exporter, "create_memory")
        assert span is not None
        assert span.instrumentation_scope is not None
        attrs = span.attributes or {}
        assert span.instrumentation_scope.name == f"{SCOPE_LATITUDE}.memory"
        assert attrs["gen_ai.operation.name"] == "create_memory"
        assert attrs["gen_ai.memory.store.id"] == "prefs"
        assert attrs["gen_ai.memory.record.id"] == "mem_1"
        assert attrs["gen_ai.memory.record.count"] == 1
        assert isinstance(attrs["gen_ai.memory.record.count"], int)
        assert "gen_ai.memory.records" not in attrs

        lat.shutdown()

    def test_sends_content_as_json_array_only_when_enabled_with_redact(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(
            lat,
            store_id="prefs",
            capture_content=True,
            redact=lambda records, _info: [{**r, "content": "[redacted]"} for r in records],
        )

        memory.update(record_id="mem_1", records=[{"id": "mem_1", "content": "likes tea", "metadata": {"src": "chat"}}])
        lat.flush()

        span = _span(exporter, "update_memory")
        assert span is not None
        raw = (span.attributes or {})["gen_ai.memory.records"]
        assert isinstance(raw, str)

        parsed: list[Any] = json.loads(raw)
        # Contract self-check: non-empty array, every element has `content`.
        assert isinstance(parsed, list)
        assert len(parsed) > 0
        assert all("content" in r for r in parsed)
        assert parsed[0]["content"] == "[redacted]"

        lat.shutdown()

    def test_wrap_sync_execute_returns_result_and_stamps_context(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(
            lat,
            store_id="prefs",
            context={"session_id": "sess-1", "user_id": "user-1", "project": "test-project"},
        )

        result = memory.create(record_id="mem_1", execute=lambda: {"ok": True})
        assert result == {"ok": True}
        lat.flush()

        span = _span(exporter, "create_memory")
        assert span is not None
        attrs = span.attributes or {}
        assert attrs["session.id"] == "sess-1"
        assert attrs["user.id"] == "user-1"
        assert attrs["latitude.project"] == "test-project"

        lat.shutdown()

    async def test_wrap_async_execute_returns_result(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(lat, store_id="prefs")

        async def do_write() -> dict[str, bool]:
            return {"ok": True}

        result = await memory.upsert(record_id="mem_1", execute=do_write)
        assert result == {"ok": True}
        lat.flush()

        assert _span(exporter, "upsert_memory") is not None
        lat.shutdown()

    async def test_sync_callable_returning_awaitable_is_awaited(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(lat, store_id="prefs")

        async def do_write() -> str:
            return "done"

        result = await memory.create(record_id="mem_1", execute=lambda: do_write())
        assert result == "done"
        lat.flush()

        assert _span(exporter, "create_memory") is not None
        lat.shutdown()

    def test_records_exception_and_reraises(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(lat, store_id="prefs")

        def boom() -> None:
            raise RuntimeError("boom")

        try:
            memory.delete(record_id="mem_1", execute=boom)
            raise AssertionError("expected RuntimeError")
        except RuntimeError as error:
            assert str(error) == "boom"
        lat.flush()

        span = _span(exporter, "delete_memory")
        assert span is not None
        assert span.status.status_code == StatusCode.ERROR
        assert any(event.name == "exception" for event in span.events)

        lat.shutdown()

    def test_search_maps_results_and_always_sets_count(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(lat, store_id="prefs")

        hits = [
            {"id": "mem_1", "content": "likes tea", "score": 0.9},
            {"id": "mem_2", "content": "lives in Barcelona", "score": 0.7},
        ]
        result = memory.search(query="preferences", execute=lambda: hits, records_from_result=lambda r: r)
        assert result == hits
        lat.flush()

        span = _span(exporter, "search_memory")
        assert span is not None
        attrs = span.attributes or {}
        assert attrs["gen_ai.memory.query.text"] == "preferences"
        assert attrs["gen_ai.memory.record.count"] == 2
        # Count is not content, so it rides even with capture off.
        assert "gen_ai.memory.records" not in attrs

        lat.shutdown()

    def test_delete_without_record_id_is_a_whole_store_wipe(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(lat)

        memory.delete(store_id="prefs")
        memory.create_store(store_id="prefs")
        lat.flush()

        wipe = _span(exporter, "delete_memory")
        assert wipe is not None
        wipe_attrs = wipe.attributes or {}
        assert wipe_attrs["gen_ai.memory.store.id"] == "prefs"
        assert "gen_ai.memory.record.id" not in wipe_attrs

        create_store = _span(exporter, "create_memory_store")
        assert create_store is not None
        assert (create_store.attributes or {})["gen_ai.operation.name"] == "create_memory_store"

        lat.shutdown()

    def test_inherits_context_from_enclosing_capture(self):
        exporter, lat = _new_latitude()
        memory = create_memory_telemetry(lat, store_id="prefs")

        capture(
            "agent-run",
            lambda: memory.create(record_id="mem_1"),
            {"session_id": "sess-9", "user_id": "user-9"},
        )
        lat.flush()

        span = _span(exporter, "create_memory")
        assert span is not None
        attrs = span.attributes or {}
        assert attrs["session.id"] == "sess-9"
        assert attrs["user.id"] == "user-9"

        lat.shutdown()
